Add cluster autoupdate resources with enabled cache

vapopov opened the pull request:
In this PR implemented new resources for autoupdate configuration and version store with enabled cache on proxy side to be able serve these resources without making requests to auth service.
Also added commands for tctl edit/get/create

Related: https://github.com/gravitational/teleport/pull/39805
changelog: Added resources for autoupdate (config and version)

vapopov committed:
Add cluster autoupdate resources with enabled cache and

vapopov committed:
Refactor autoupdate service from legacy approach

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/add-autoupdate-resources

vapopov committed:
Add commands for autoupdate

vapopov committed:
Add test and fix CI

vapopov committed:
Rename autoupdate config resource

vapopov committed:
Add cache test and comment fixes

vapopov commented on api/proto/teleport/autoupdate/v1/autoupdate.proto:
```diff
@@ -0,0 +1,55 @@
+// AutoupdateConfigSpec encodes the parameters of the autoupdate config object.
+message AutoupdateConfigSpec {
+  // ToolsAutoUpdate encodes the feature flag to enable/disable tools autoupdates.
+  bool tools_auto_update = 1;
```
I would need to change in client RFD as well

vapopov replied:
✅ https://github.com/gravitational/teleport/pull/39805/commits/e4f4499dadf9b3ec249a37ce0f746a1363af0947

vapopov committed:
Rename tools_autoupdate

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/add-autoupdate-resources

vapopov committed:
Fix events from latest change in backend lib

sclevine reviewed (approved):
LGTM, may want to double-check with @rosstimothy about whether the cache is configured suitably for serving the /find endpoint

sclevine commented on tool/tctl/common/autoupdate_command.go:
```diff
@@ -0,0 +1,206 @@
+	autoupdateCmd := app.Command("autoupdate", "Teleport autoupdate commands.")
+
+	c.updateCmd = autoupdateCmd.Command("update", "Edit autoupdate configuration.")
+	c.updateCmd.Flag("set-tools-auto-update", `Enable or disable tools autoupdate in cluster.`).EnumVar(&c.toolsAutoupdate, "on", "off")
```
```suggestion
	c.updateCmd.Flag("set-tools-autoupdate", `Enable or disable tools autoupdate in cluster.`).EnumVar(&c.toolsAutoupdate, "on", "off")
```
Feel free to ignore, but if we want to be super-consistent about it, could change

vapopov replied:
missed this one, will update with rfd

sclevine commented on tool/tctl/common/autoupdate_command.go:
```diff
@@ -0,0 +1,206 @@
+	current := teleport.SemVersion
+	ticker := interval.New(interval.Config{
+		Duration:      time.Minute,
+		FirstDuration: time.Second,
```
Curious why 1s instead of 0?

vapopov replied:
```go
	firstDuration := cfg.FirstDuration
	if firstDuration == 0 {
		firstDuration = interval.duration()
	}
```
it is going to be ignored if there are 0, made similar to some places where interval lib is used
I would probably change to this one:

```go
	ticker := interval.New(interval.Config{
		Duration:      0,
		FirstDuration: time.Second,
	})
	ticker.FireNow()
	defer ticker.Stop()
```
also I forgot to add `defer ticker.Stop()`

sclevine replied:
Also option to structure it like this:
```golang
for {
  c.get()
  case ...
}

vapopov committed:
Command flag naming adjustment
Fix interval first duration

rosstimothy reviewed (commented):
There is a lot going on in this PR - many layers of boilerplate all being implemented at once. I'd suggest trying to scope this to just adding the resource, the RPC service and persisting it in the backend and retrieving it from the cache. The tctl commands and webapi changes could probably both be in their own distinct and more focused PRs.

sclevine commented on tool/tctl/common/autoupdate_command.go:
```diff
@@ -0,0 +1,205 @@
+	ToolsVersion string `json:"tools_version"`
+}
+
+// AutoupdateCommand implements the `tctl autoupdate` command for managing
+// autoupdate process for tools and agents.
+type AutoupdateCommand struct {
```
The `autoupdate` subcommand was proposed by RJ here: https://github.com/gravitational/teleport/pull/39805/files#diff-8d3c0408dd826f27e9f13cd15f09f4d7e89202862f0b13ce131a2e3072c2a40dR249

Seems reasonable for both the subcommand and `tctl create`/`tctl get` to work?

vapopov replied:
I believe that might be better from UX, when everything in one place and you able to configure autoupdates without having knowledge about resources responsible for this configuration.
I will additionally add these resources also to `tctl create/get` resource command

vapopov commented on lib/auth/authclient/clt.go:
```diff
@@ -360,6 +361,44 @@ func (c *Client) SearchSessionEvents(ctx context.Context, req events.SearchSessi
 }
 
+// UpsertAutoupdateConfig sets autoupdate configuration.
+func (c *Client) UpsertAutoupdateConfig(ctx context.Context, config *autoupdate.AutoupdateConfig) (*autoupdate.AutoupdateConfig, error) {
+	return nil, trace.NotImplemented(notImplementedMessage)
+}
+
+// GetAutoupdateConfig gets the autoupdate configuration from the backend.
+func (c *Client) GetAutoupdateConfig(ctx context.Context) (*autoupdate.AutoupdateConfig, error) {
+	config, err := c.APIClient.GetAutoupdateConfig(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return config, nil
+}
+
+// DeleteAutoupdateVersion deletes types.AutoupdateVersion from the backend.
+func (c *Client) DeleteAutoupdateVersion(ctx context.Context) error {
+	return trace.NotImplemented(notImplementedMessage)
+}
+
+// UpsertAutoupdateVersion sets autoupdate version.
+func (c *Client) UpsertAutoupdateVersion(ctx context.Context, config *autoupdate.AutoupdateVersion) (*autoupdate.AutoupdateVersion, error) {
+	return nil, trace.NotImplemented(notImplementedMessage)
+}
+
+// GetAutoupdateVersion gets the autoupdate version from the backend.
+func (c *Client) GetAutoupdateVersion(ctx context.Context) (*autoupdate.AutoupdateVersion, error) {
+	config, err := c.APIClient.GetAutoupdateVersion(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return config, nil
+}
+
+// DeleteAutoupdateConfig deletes types.AutoupdateConfig from the backend.
+func (c *Client) DeleteAutoupdateConfig(ctx context.Context) error {
+	return trace.NotImplemented(notImplementedMessage)
+}
+
```
we don't need this one, I've removed this part

vapopov committed:
Simplify client implementation for autoupdate

vapopov commented on api/client/client.go:
```diff
@@ -2863,6 +2869,24 @@ func (c *Client) GetClusterAuditConfig(ctx context.Context) (types.ClusterAuditC
 	return resp, nil
 }
 
+// GetAutoupdateConfig gets autoupdate configuration.
+func (c *Client) GetAutoupdateConfig(ctx context.Context) (*autoupdate.AutoupdateConfig, error) {
+	resp, err := c.AutoupdateServiceClient().GetAutoupdateConfig(ctx, &autoupdate.GetAutoupdateConfigRequest{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return resp, nil
+}
+
+// GetAutoupdateVersion gets autoupdate version.
+func (c *Client) GetAutoupdateVersion(ctx context.Context) (*autoupdate.AutoupdateVersion, error) {
+	resp, err := c.AutoupdateServiceClient().GetAutoupdateVersion(ctx, &autoupdate.GetAutoupdateVersionRequest{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return resp, nil
+}
```
after adding get methods for fetching config and version in `find` endpoint
```golang
interface ReadRemoteProxyAccessPoint {
	GetAutoupdateConfig(ctx context.Context) (*autoupdate.AutoupdateConfig, error)
	GetAutoupdateVersion(ctx context.Context) (*autoupdate.AutoupdateVersion, error)
}
```
`cache.Cache` and `client.ClientI` should implement these methods, otherwise if I replace this one to client get method I need to add some wrapper for https://github.com/gravitational/teleport/pull/45617/files#diff-3e237f973a30f7092dc18a706ecd95d8a7e1d99c5c50626b8bb0e7ac6cb7efd1R1899-R1920

vapopov committed:
Add create/update api for autoupdate service
Add resource command for autoupdate config and version
Limit to getter interface for client and cache part
Add client wrapping

vapopov committed:
Add tests for service and resource validation

vapopov commented on lib/auth/autoupdate/v1/service.go:
```diff
@@ -0,0 +1,148 @@
+
+// GetAutoupdateConfig gets the current autoupdate config singleton.
+func (s *Service) GetAutoupdateConfig(ctx context.Context, req *autoupdate.GetAutoupdateConfigRequest) (*autoupdate.AutoupdateConfig, error) {
+	config, err := s.storage.GetAutoupdateConfig(ctx)
```
we use cache wrapper on top of auth client (if cache enabled on proxy side), or do you mean we should have cache on service side, similar to clusterconfig?

vapopov committed:
Drop autoupdate tctl command and find endpoint changes

vapopov committed:
Fix tests

vapopov committed:
Re-use cache for autoupdate service
Add tests for resources command

vapopov commented:
@rosstimothy could you please take another look at the PR?

rosstimothy reviewed (commented):
As mentioned before, I think it would be best if this PR was split up into several smaller, more focused PRs. I tend to recommend that for new resources folks try to split each layer of the boilerplate up into PRs in roughly the following order:

1) resource and audit events proto and generated code
2) backend storage service implementation
3) RPC service implementation
4) cache support
5) tctl support
6) audit event support


A good example to follow is the [`SPIFFEFederation`]( https://github.com/gravitational/teleport/pulls?q=is%3Apr+SPIFFEFederation+is%3Aclosed) resource that Noah recently added.

vapopov commented on api/types/constants.go:
```diff
@@ -320,6 +320,18 @@ const (
 	// alternative to their individual resource kinds.
 	KindClusterConfig = "cluster_config"
 
+	// KindAutoupdateConfig is the resource with autoupdate configuration.
+	KindAutoupdateConfig = "autoupdate_config"
+
+	// KindAutoupdateVersion is the resource with autoupdate versions.
+	KindAutoupdateVersion = "autoupdate_version"
+
+	// MetaNameAutoupdateConfig is the name of a configuration resource for autoupdate config.
+	MetaNameAutoupdateConfig = "autoupdate-config"
+
+	// MetaNameAutoupdateVersion is the name of a resource for autoupdate version.
+	MetaNameAutoupdateVersion = "autoupdate-version"
```
We agreed with @sclevine to use `Autoupdate` instead camel case, both RFDs was modified to be consistent

sclevine replied:
No strong opinion about this. The RFDs had a mixture of `autoupdate_` / `Autoupdate` and `auto_update_` / `AutoUpdate` prefixes, so we went with the former spelling everywhere to be consistent with the resource and command naming in the original Client Tools proposal: https://github.com/gravitational/teleport/pull/39805/files

@rosstimothy if you feel strongly that we should move the other way, or break consistency for identifiers in the codebase, no objections.

vapopov commented on api/client/autoupdate/autoupdate.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+// Client is an AutoupdateService client that conforms to the following lib/services interfaces:
+//   - services.AutoupdateService
```
ok, will remove this one, we have other 4 clients declaring the same

vapopov commented on api/client/autoupdate/autoupdate.go:
```diff
@@ -0,0 +1,96 @@
+
+// Client is an AutoupdateService client that conforms to the following lib/services interfaces:
+//   - services.AutoupdateService
+type Client struct {
```
in codebase there are 3 ways how it is implemented, legacy one with interface in types, methods implementation and this one with wrapping grpc client to be interface compatible to service definition, this PR by iterations implemented all of them.

Last one was added because I was trying to eliminate methods declarations for client (as it was requested previously in this review)
https://github.com/gravitational/teleport/commit/187769517a480b738f2db5c2c2d9735d9d5665e5#diff-f5e514e352c12c01eab507e6d9bbdfa697791b35bf914d949b92245a8e950447R2873-R2889

and use only client wrapper:
https://github.com/gravitational/teleport/commit/187769517a480b738f2db5c2c2d9735d9d5665e5#diff-3bdff46f360deb2937ae725b5117a2462cf0bcaaff86400d08179c652a2ce2f3R2653

but its not possible right now because if we use cache wrapping, interfaces of the `cache.Cache` and `authclient.ClientI` must be compatible

```golang
interface ReadRemoteProxyAccessPoint {
	GetAutoupdateConfig(ctx context.Context) (*autoupdate.AutoupdateConfig, error)
	GetAutoupdateVersion(ctx context.Context) (*autoupdate.AutoupdateVersion, error)
}
type RemoteProxyAccessPoint interface {
	ReadRemoteProxyAccessPoint
	accessPoint
}

func NewRemoteProxyWrapper(base RemoteProxyAccessPoint, cache ReadRemoteProxyAccessPoint) RemoteProxyAccessPoint {...}

// newLocalCacheForRemoteProxy returns new instance of access point configured for a remote proxy.
func (process *TeleportProcess) newLocalCacheForRemoteProxy(clt authclient.ClientI, cacheName []string) (authclient.RemoteProxyAccessPoint, error) {
	// if caching is disabled, return access point
	if !process.Config.CachePolicy.Enabled {
		return clt, nil
	}

	cache, err := process.NewLocalCache(clt, cache.ForRemoteProxy, cacheName)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	return authclient.NewRemoteProxyWrapper(clt, cache), nil
}
```

vapopov commented:
closed in prior to several PRs described in https://github.com/gravitational/teleport/pull/45617#pullrequestreview-2274490029
